packages/doenetml-worker-rust/lib-doenetml-core/src/core/dast/macro_resolve.rs

```diff
@@ -0,0 +1,230 @@
+    /// - Matching `x` from anywhere returns the index of `<a />`.
+    /// - Matching `y` from `<c />` returns the index of `<b />`.
+    /// - Matching `y` from `<d />` returns an error because there are multiple `y`s accessible from that position.
+    /// - Matching `y.z.w` from `<b />` returns the index of `<c />` along with `.w` as the unresolved path.
+    ///
+    /// If an index appears in the path (e.g. `a[2].b`), the search stops and the remaining path is returned.
+    /// If there was a partial match of the indexed item, the unresolved path will list `name` as an empty string.
+    /// E.g., matching `y.w[2]` from `<b />` returns the index of `<d />` along with `.w[2]` as the unresolved path
+    /// and matching `y[2]` from `<b />` returns the index of `<d />` along with `.[2]` as the unresolved path.
+    pub fn resolve<T: AsRef<[PathPart]>>(
+        &self,
+        path: T,
+        origin: Index,
+    ) -> Result<RefResolution, ResolutionError> {
+        let path = path.as_ref();
+        let mut current_idx = origin;
+        let mut path = path.iter();
+        let first_path_part = path.next().ok_or(ResolutionError::NoReferent)?;
+
+        current_idx = self.search_parents(&first_path_part.name, current_idx)?;
+        // If we made it here, the first entry in `path` has a valid referent.
+        // If this entry also has an index, we need to stop searching. This would
+        // happen if the reference was something like `$a[2].b`.
+        if !first_path_part.index.is_empty() {
+            let remaining_path: Vec<PathPart> = iter::once(PathPart {
+                name: "".into(),
+                index: first_path_part.index.clone(),
+                position: first_path_part.position.clone(),
+            })
+            .chain(path.cloned())
+            .collect();
+            return Ok(RefResolution {
+                node_idx: current_idx,
+                unresolved_path: Some(remaining_path),
+            });
+        }
+
+        while let Some(part) = path.next() {
+            // current_idx specifies the "root" of the search. We try to resolve
+            // children based on the path part, breaking if any of them are not found or if
+            // there is an `index` specified.
+            if let Some(referent) = self.name_map[current_idx].get(&part.name) {
+                match referent {
+                    Ref::Unique(idx) => {
+                        current_idx = *idx;
+                    }
+                    Ref::Ambiguous(_) => {
+                        return Err(ResolutionError::NonUniqueReferent);
+                    }
+                }
+                // If there index specified, we immediately stop since component information is needed
+                // to resolve all remaining path parts.
+                if !part.index.is_empty() {
+                    let remaining_path: Vec<PathPart> =
+                        iter::once(part.clone()).chain(path.cloned()).collect();
```

Comment:
So we have different behavior for `$x[0]` and `$x.y[0]`? For `$x[0]`, we match `x` and have remaining path be `""[0]`.  For `$x.y[0]`, it seems we don't match `y` but just match `x` and have the remaining path be `y[0]`. Is there a reason we don't match `x.y` and have the remaining path part be `""[0]`. That seems more consistent to me, but I'm not confident that I'm not missing something.

Reply:
The reason is me not thinking clearly :-P. I even hardcoded this in a test...but I think the behavior should be consistent (especially if we allow indexing of items directly rather than just indexing of state variables)

Reply:
Fixed, and a test is added.